### PR TITLE
statistics ui created

### PR DESCRIPTION
**Pull request by HySpeed**
* added stats ui and event handling
* todo list updated

### CODE REVIEW

File: scripts/coe_player.lua [hunk 1]
```diff
@@ -4,6 +4,7 @@ local Player = {}
 local Config = require("config")
 local Log = require("utils/log")
 local Utils = require("utils/utils.lua")
+local Stats = require("scripts/coe_stats_gui")
```
Comment (Nexela) [inline scripts/coe_player.lua:7]: Player is a helper class, it shouldn't require anything. coe_stats_gui needs an onPlayerCreated event and have it call the setupStatsGui function

Then add StatsGui.OnPlayerCreated() to the event in control.lua

File: control.lua [hunk 2]
```diff
@@ -82,7 +81,13 @@ script.on_event(defines.events.on_gui_opened, TeleporterGUI.onGuiOpened)
 
 script.on_event(defines.events.on_gui_closed, TeleporterGUI.onGuiClosed)
 
-script.on_event(defines.events.on_gui_click, TeleporterGUI.onGuiClick)
+script.on_event(defines.events.on_gui_click, function(event)
+  if event.element.name == "coe_teleporter_gui_close" or next( event.element.tags ) ~= nil then
```
Comment (Nexela) [inline control.lua:85]: The event handler in each file should handle this logic. I believe it is already handled for Teleporter
Also the stats_gui script has not been added

File: todo.lua [hunk 3]
```diff
@@ -0,0 +1,6 @@
+
+-- TODO: add "discharge equipment" after teleport
+-- TODO: When the setting "All teleporters available" is true, then the Teleporter GUI should show all Cities.
+-- TODO: Remove the map tag 'ping' when a teleport happens. (this will be annoying and spammy in a multiplayer game.)
```
Comment (Nexela) [inline todo.lua:4]: Already Done

File: control.lua (hunk 2 above)
Comment (Nexela) [reply control.lua:85]: Additionaly I think its time to add simple event handler!

**Review by Nexela — CHANGES_REQUESTED** [review]

File: scripts/coe_stats_gui.lua [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      type = "frame",
+      name = "coe_statistics_frame",
+      direction = "vertical",
+      caption = {"coe-stats-gui.label_stats_title"}
+  }
+
+  build_stats_info_frame( statistics_frame )
+
+  local pdata = global.players[event.player_index]
+  pdata.button_statistics_visible = true
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:55]: Whats the reason for storing the visible state of this window?

File: scripts/coe_stats_gui.lua [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    type = "label",
+    caption = tostring( global.rocket_silo.total_launches )
+  }
+end -- build_stats_info_frame
+
+-------------------------------------------------------------------------------
+
+local function open_stats_ui(event)
+  local player = game.get_player(event.player_index)
+  local gui = player.gui.left
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:43]: Statistics frame should open under the mod button

File: scripts/coe_stats_gui.lua [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  })
+end -- setupStatsGUI
+
+-------------------------------------------------------------------------------
+
+---@param event EventData.on_gui_click
+function StatsGUI.onGuiClick(event)
+  if event.element.name == "coe_button_statistics" then
+    local pdata = global.players[event.player_index]
+    if pdata.button_statistics_visible then
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:94]: Check for the presence of the window and destroy it then recreate it.

File: scripts/coe_stats_gui.lua [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      player.gui.left.coe_statistics_frame.destroy()
+    end
+  end
+  local pdata = global.players[event.player_index]
+  pdata.button_statistics_visible = false
+end -- close_stats_gui
+
+-------------------------------------------------------------------------------
+
+function StatsGUI.setupStatsGUI( event )
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:74]: Rename to OnPlayerCreated

File: scripts/coe_stats_gui.lua [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  local player = game.get_player(event.player_index)
+  local dialog = mod_gui.get_button_flow( player )
+  if dialog.coe_button_statistics then dialog.coe_button_statistics.destroy() end
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:77]: Unneeded check

File: scripts/coe_stats_gui.lua [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  build_stats_info_frame( statistics_frame )
+
+  local pdata = global.players[event.player_index]
+  pdata.button_statistics_visible = true
+
+end -- open_stats_ui
+
+-------------------------------------------------------------------------------
+
+local function close_stats_gui( event )
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:61]: event param needed here?

File: scripts/coe_stats_gui.lua [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    type = "flow",
+    direction = "horizontal"
+  }.add{
+    type = "label",
+    caption = {"coe-stats-gui.label_launches_complete"}
+  }.parent.add{
+    type = "label",
+    caption = tostring( global.rocket_silo.total_launches )
+  }
+end -- build_stats_info_frame
```
Comment (Nexela) [inline scripts/coe_stats_gui.lua:37]: function should return the info_frame

**Review by Nexela — CHANGES_REQUESTED** [review]

File: control.lua (hunk 2 above)
Comment (HySpeed) [reply control.lua:85]: So control.lua can have multiple `script.on_event( ... )` for the same event?

File: scripts/coe_stats_gui.lua (hunk 4 above)
Comment (HySpeed) [reply scripts/coe_stats_gui.lua:55]: using a tracking variable tied to the player is how the UI tutorial showed how to know if the UI was open.
I don't know if it is right.  I'll explore checking the UI elements directly.

File: scripts/coe_stats_gui.lua (hunk 5 above)
Comment (HySpeed) [reply scripts/coe_stats_gui.lua:43]: I do not understand this request.

File: scripts/coe_stats_gui.lua (hunk 4 above)
Comment (Nexela) [reply scripts/coe_stats_gui.lua:55]: In some instance a tracking variable can be a good thing, but for a small on click gui that is created and destroyed I wouldnt bloat the global with it

File: control.lua (hunk 2 above)
Comment (Nexela) [reply control.lua:84]: Only one per event,  but each of our script files can have an onXXX handler that we call inside the script.on_event

**Pull request closed without merging**